site/docs/tutorials/kafka.md

```diff
@@ -3,10 +3,253 @@ title: Work with Apache Kafka
+ .withoutTimestamps()
+ .writeTo(Sinks.map("cacheMap"));
+
+JetInstance jet = Jet.bootstrappedInstance();
+Job job = jet.newJob(p);
+```
+
+## Complex Record Types
+
+You can put your complex objects to Kafka using custom serializers and
+de-serializers. Jet expects the consumed objects to be `Serializable`
+since the objects will be moved between stages. You can also define a
+projection function for source which will be applied to each record
+before emitting to downstream. You can also define key and value
+extractors for sink. Let's rewrite enrich data sample using a complex
+object as the value of the record.
+
+```java
+Properties properties = new Properties();
+properties.setProperty("bootstrap.servers", "localhost:9092");
+properties.setProperty("key.deserializer", StringDeserializer.class.getCanonicalName());
+properties.setProperty("value.deserializer", UserDeserializer.class.getCanonicalName());
+properties.setProperty("key.serializer", StringSerializer.class.getCanonicalName());
+properties.setProperty("value.serializer", UserSerializer.class.getCanonicalName());
+
+
+Pipeline p = Pipeline.create();
+p.readFrom(KafkaSources.kafka(properties, ConsumerRecord<String, User>::value, "sourceTopic"))
+ .withoutTimestamps()
+ .mapUsingIMap("passwordMap", User::getName, (user, pass) -> user.setPassword((byte[])pass))
+ .writeTo(KafkaSinks.kafka(properties, "sinkTopic", User::getName, user -> user));
+
+JetInstance jet = Jet.bootstrappedInstance();
+Job job = jet.newJob(p);
+```
+
+The **User** class:
+
+```java
+public class User implements Serializable {
+
+    private String name;
+    private byte[] password;
+
+    public User() {
+    }
+
+    public User(String name, byte[] password) {
+        this.name = name;
+        this.password = password;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public User setName(String name) {
+        this.name = name;
+        return this;
+    }
+
+    public byte[] getPassword() {
+        return password;
+    }
+
+    public User setPassword(byte[] password) {
+        this.password = password;
+        return this;
+    }
+}
+```
+
+## Schema Registry
+
+While you can use custom serializer/de-serializer for your complex
+objects, you can store your object as `json` or `avro` and use a **Schme
+Registry** to handle the metadata.
+
+```java
+Pipeline p = Pipeline.create();
+
+Properties properties = new Properties();
+properties.setProperty("bootstrap.servers", "localhost:9092");
+properties.setProperty("key.deserializer", StringDeserializer.class.getCanonicalName());
+properties.setProperty("value.deserializer", KafkaAvroDeserializer.class.getCanonicalName());
+properties.setProperty("schema.registry.url", "http://localhost:8081");
+properties.setProperty("specific.avro.reader", "true");
+
+p.readFrom(KafkaSources.<String, User>kafka(properties, "sourceTopic"))
+ .withoutTimestamps()
+ .writeTo(Sinks.logger());
+```
+
+The **User** class with avro schema:
```

Comment:
I think it's easier to use an auto-generated one, since you upload the schema file to registry normally. I will merge this for now and you can update later.